### ISSUE TITLE
Clickable area on dropdown menu

### ISSUE THREAD

**Issue by brianteeman** [joomla/cassiopeia#125]
### Steps to reproduce the issue
Install current development branch and sample data

### Expected result
When you hover on a link and it is visually indicated as a link (in this case by a line) then the link is clickable

### Actual result
Only the icon is clickable

![menu](https://user-images.githubusercontent.com/1296369/94200242-03d2c580-feb2-11ea-9468-7b655c3f925e.gif)

and on mobile its worse as there is no link line and the dropdown that is clickable is a long way from the text

![mobile](https://user-images.githubusercontent.com/1296369/94200386-40062600-feb2-11ea-9b6a-99735482f44d.gif)

**Comment by richard67** [comment]
The "Sample Layouts" should not be a link because it is only a heading. The "Articles" in the dropdown, which is also a parent, is a "real" (i.e. not heading or separator) menu item with a link. The mouse pointer shows if it's a link or not.

Not trying to justify, only trying to explan what we have.

**Comment by richard67** [comment]
So the underline might be wrong for the "Sample Layouts".

**Comment by brianteeman** [comment]
Thats even worse - the heading should be clickable if its the way you expand the menu

**Comment by richard67** [comment]
The menu should be only expanded with the triangle.

**Comment by brianteeman** [comment]
Then that is WRONG, VERY WRONG

**Comment by richard67** [comment]
Well, then it has to be corrected.

**Comment by brianteeman** [comment]
See the top menu bar at the metismenu example site
https://mm.onokumus.com/

**Comment by richard67** [comment]
@brianteeman Yes, that's for a heading. For a menu item which is parent of a dropdown and has a link, like e.g. the "Articles" in the "Sample Layouts" (something which we can't find in metimenu examples), is it ok like we do it there, the "Articles" is the link, and the triangle opens the dropdown? Or is there something wrong? Or should we not do that at all, have dropdown parents with an own link?

**Comment by brianteeman** [comment]
No it is not ok. If nothing else it is unexpected behaviour. As I am not aware of other sites that do this. 

From an accessibility point of view it also an even bigger fail.

When a screenreader comes across the menu it will say in sequence

Link - blog
link - about
link - author login
toggle dropdown - button - collapsed

So if all the menus were dropdowns it would be
toggle dropdown - button - collapsed
toggle dropdown - button - collapsed
toggle dropdown - button - collapsed
toggle dropdown - button - collapsed
toggle dropdown - button - collapsed
toggle dropdown - button - collapsed
toggle dropdown - button - collapsed
toggle dropdown - button - collapsed

**Comment by brianteeman** [comment]
Super quick proof of concept

![menu](https://user-images.githubusercontent.com/1296369/94208237-46e86500-fec1-11ea-9a7e-1adbe83ee0e4.gif)
![chrome_2020-09-24_23-49-22](https://user-images.githubusercontent.com/1296369/94208125-07217d80-fec1-11ea-8806-c4828103bfb6.png)
![chrome_2020-09-24_23-49-36](https://user-images.githubusercontent.com/1296369/94208129-07ba1400-fec1-11ea-841a-f44ce7b43288.png)

**Comment by chmst** [comment]
The solution is as we had in the beginning and in the backend menu. Headings open the submenu but are no links.

**Comment by richard67** [comment]
@chmst Yes, but when we had extended this I asked if this is accessible or not, and nobody said it's not. I have no problem with it when we change it back.

**Comment by chmst** [comment]
We have discussed that and had the information that tihs is accessible. It is all right for keyboard users, every element is focusable.  

In my opinion it is not userfriendly but was required for the dropdown - B/C.

**Comment by brianteeman** [comment]
Try it with a screenreader https://github.com/joomla/cassiopeia/issues/125#issuecomment-698602906

Thats certainly not accessible

**Comment by brianteeman** [comment]
I can get close to fixing the clickable area (not the accessibility) by changing 
https://github.com/joomla/cassiopeia/blob/be0c5669142f4f456f5e2c550ff92268dbf2a553/templates/cassiopeia/html/mod_menu/metismenu.php#L100

to include the class stretched-link

and by removing position: relative from
`.metismenu.mod-menu .mm-toggler` 

But then my css skills fail me

**Comment by chmst** [comment]
Most important is accessibiliy. This must be an accessible drop down menu. Parent Item menus will open the submenu, as we have in the backend.

**Comment by brianteeman** [comment]
@chmst I agree with you of course. But if you check the admin you will see that we do not use menu items of type header in the admin at all.

**Comment by richard67** [comment]
I'm working on a possible solution: #184 ... but it's still work in progress and needs at least some more scss changes for not having the item underlined when hovering if it doesn't have a link.

**Comment by richard67** [comment]
@brianteeman I need your advise regarding this issue while working on a fix.

You wrote in the issue description as expected result:

> When you hover on a link and it is visually indicated as a link (in this case by a line) then the link is clickable

My question: Let's assume I have fixed this kind of menu item so that the complete thing is clickable, but it's a button and not a link, should it be underlined when hovering because it's clickable? Or should this only be done for real hyperlinks but not for buttons?

Thanks in advance for the answer.

**Comment by brianteeman** [comment]
Well it shouldnt be a button

**Comment by richard67** [comment]
In case if it is a menu item of type heading or separator there is no `<a>`.

Assuming the complete menu item will be the dropdown toggle and I shall not use a button, shall I use `<a href="#">` instead?

If so: Should it have a `role="button"` then?

**Comment by richard67** [comment]
You can check the description of #184 for what I'm currently doing, including markup. Feel free to make comments there if that's the better place.

**Comment by brianteeman** [comment]
I will take a look tomorrow morning

**Comment by brianteeman** [comment]
Can you check the menu I have at https://sinaileeds.uk

The first lot of top menu items are alias and the one labelled HHD is a heading

**Comment by richard67** [comment]
@brianteeman As I can see the alias menu items have a link, but when clicking that item it is only a dropdown toggle. In the dropdown then is at the top a menu item with a different title and the same link. This is of course a way to do it. But it means you have a menu item, e.g. "About", where the link is not really used, and another one, "About Sinai", where it is used, and you have to care yourself for not having duplicate titles, which would be confusing.

I still think it should be possible with what bootstrap 4 calls a split button, with link and dropdown, both working, like I try to do it in my PR. If that PR is not completely wrong with that, I can easily change the buttons to links with href="#", that would even make styling much easier.

If then someone wants to have the same structure like in the menu of the site you have linked, he or she still can do that by using a heading or alias for the dropdown parent and then in the dropdown having another menu item with a slightly different title but the same link, like it can be done now already.

**Comment by brianteeman** [comment]
The hhd menu is a header not an alias

**Comment by richard67** [comment]
> The hhd menu is a header not an alias

Yes, I saw. I talked about the others.

**Comment by brianteeman** [comment]
For the purpose of the pr I just wanted to show that it can be done with a
heading

**Comment by richard67** [comment]
I see it is an `<a>` without any `href` property, not even `href="#"`.

```
<a class="nav-header nav-link dropdown-toggle" data-toggle="dropdown">HHD</a>
```

As I've learned just now, reading specs, it is allowed in HTML5 to have that, `<a>` without any `href`. Before HTML5 it was required.

Shall I do it in that way in my PR, replacing the buttons with `<a>` elements?

**Comment by brianteeman** [comment]
It is a suggestion. I am not telling anyone to do anything especially regarding accessibility

**Comment by brianteeman** [comment]
BTW with this suggestion there is no need for metismenu at all

**Comment by richard67** [comment]
I've decided to replace the PR #184 by a new one with identical changes but using a branch in this repository here instead of my fork. This makes testing and contributing easier.

New PR is #195 .

@brianteeman I'd still be happy if you could have a look on it.

**Comment by richard67** [comment]
Solved with pull request #195 .
